src/python/pants/backend/project_info/peek.py

```diff
@@ -0,0 +1,146 @@
+            "--output",
+            type=OutputOptions,
+            default=OutputOptions.RAW,
+            help="which output style peek should use",
+        )
+
+    @property
+    def output(self) -> OutputOptions:
+        return cast(OutputOptions, self.options.output)
+
+
+class Peek(Goal):
+    subsystem_cls = PeekSubsystem
+
+
+_T = TypeVar("_T")
+_U = TypeVar("_U")
+
+
+def _intersperse(x: _T, ys: Iterable[_U]) -> Iterable[Union[_T, _U]]:
+    # Use stateful iterator to safely iterate through first item
+    it = iter(ys)
+    for y in it:
+        yield y
+        break
+    for x, y in zip(repeat(x), it):
+        yield x
+        yield y
+
+
+def _render_raw(fc: FileContent, encoding: str = "utf-8") -> str:
+    dashes = "-" * len(fc.path)
+    content = fc.content.decode(encoding)
+    parts = [dashes, fc.path, dashes, content]
+    if not content.endswith(os.linesep):
+        parts.append("")
+    return os.linesep.join(parts)
+
+
+def _render_json(ts: Iterable[Target]) -> str:
+    data = dict(map(_target_to_kv, ts))
+    return json.dumps(data, indent=2, cls=_PeekJsonEncoder)
+
+
+def _target_to_kv(t: Target) -> Tuple[str, dict]:
+    nothing = object()
+    d = {
+        k.alias: v.value
+        for k, v in t.field_values.items()
+        # don't report default values in normalized output
+        if getattr(k, "default", nothing) != v.value
+    }
+    d["alias"] = t.alias
+    return t.address.spec, d
+
+
+class _PeekJsonEncoder(json.JSONEncoder):
+    """Allow us to serialize some commmonly-found types in BUILD files."""
+
+    safe_to_str_types = (Requirement,)
+
+    def default(self, o):
+        """Return a serializable object for o."""
+        if is_dataclass(o):
+            return asdict(o)
+        elif isinstance(o, Mapping):
+            return dict(o)
+        elif isinstance(o, self.safe_to_str_types):
+            return str(o)
+        else:
+            try:
+                return super().default(o)
+            except TypeError:
+                # punt and just try str()
+                # TODO: can we find a better strategy here?
+                return str(o)
+
+
+@goal_rule
+async def peek(
+    console: Console,
+    subsys: PeekSubsystem,
+    addresses: Addresses,
+) -> Peek:
+    # TODO: better options for target selection (globs, transitive, etc)?
+    targets: Iterable[Target]
+    # TODO: handle target not found exceptions better here
+    targets = await Get(Targets, Addresses, addresses)
+
+    if subsys.output == OutputOptions.RAW:
+        build_file_addresses = await MultiGet(
+            Get(BuildFileAddress, Address, t.address) for t in targets
+        )
+        build_file_paths = {a.rel_path for a in build_file_addresses}
+        digest_contents = await Get(DigestContents, PathGlobs(build_file_paths))
+        # TODO: programmatically determine correct encoding
+        rendereds = map(_render_raw, digest_contents)
+        for output in _intersperse(os.linesep, rendereds):
```

Comment:
I'm not following what's happening with the intersperse method? Would it work to instead do this?

```python
for file_content in digest_contents:
  console.print_stdout(_render_raw(file_content))
```

Reply:
Regarding ordering, I think we likely would want to use alphabetical ordering based on the `file_content.path`.

Reply:
I was trying to get too cute with `intersperse`, I'll simplify.

- [ ] remove `_intersperse()`
- [ ] alphabetically order along paths